core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1899,60 +1826,53 @@ class KafkaApis(val requestChannel: RequestChannel,
-                val filter = deleteAclsRequest.filters().get(i)
-                if (filter.matches(binding)) {
-                  toDelete.get(i).get += ((resource, acl))
-                }
-              }
+        val filters = deleteAclsRequest.filters.asScala
+        val filterResponseMap = mutable.Map[Int, AclFilterResponse]()
+        val toDelete = mutable.Map[Int, ArrayBuffer[(Resource, Acl)]]()
+
+        if (filters.forall(_.matchesAtMostOne)) {
+          // Delete based on a list of ACL fixtures.
+          for ((filter, i) <- filters.zipWithIndex) {
+            SecurityUtils.convertToResourceAndAcl(filter) match {
+              case Failure(throwable) => filterResponseMap.put(i, new AclFilterResponse(throwable, Seq.empty.asJava))
+              case Success(fixture) => toDelete.put(i, ArrayBuffer(fixture))
             }
           }
         } else {
-          // Delete based on a list of ACL fixtures.
-          for (i <- 0 to deleteAclsRequest.filters().size - 1) {
-            toScala(deleteAclsRequest.filters().get(i)) match {
-              case Failure(throwable) => filterResponseMap.put(i,
-                new AclFilterResponse(throwable, Collections.emptySet[AclDeletionResult]()))
-              case Success(fixture) => toDelete.put(i, ListBuffer(fixture))
-            }
+          // Delete based on filters that may match more than one ACL.
+          val aclMap = auth.getAcls()
+          val filtersWithIndex = filters.zipWithIndex
+          for ((resource, acls) <- aclMap; acl <- acls) {
+            val binding = new AclBinding(
+              new AdminResource(AdminResourceType.fromString(resource.resourceType.toString), resource.name),
+              new AccessControlEntry(acl.principal.toString, acl.host.toString, acl.operation.toJava,
+                acl.permissionType.toJava))
+
+            for ((filter, i) <- filtersWithIndex if filter.matches(binding))
+              toDelete.getOrElseUpdate(i, ArrayBuffer.empty) += ((resource, acl))
           }
         }
-        for (i <- toDelete.keys) {
-          val deletionResults = new util.ArrayList[AclDeletionResult]()
-          for (acls <- toDelete.get(i)) {
-            for ((resource, acl) <- acls) {
-              try {
-                if (auth.removeAcls(immutable.Set(acl), resource)) {
-                  deletionResults.add(new AclDeletionResult(null, toJava((resource, acl))))
-                }
-              } catch {
-                case throwable: Throwable => deletionResults.add(new AclDeletionResult(
-                  new UnknownServerException("Failed to delete ACL: " + throwable.toString),
-                    toJava((resource, acl))))
-              }
+
+        for ((i, acls) <- toDelete) {
+          val deletionResults = acls.flatMap { case (resource, acl) =>
+            val aclBinding = SecurityUtils.convertToAclBinding(resource, acl)
+            try {
+              if (auth.removeAcls(immutable.Set(acl), resource))
```

Comment:
The Authorizer interface has complicated (and I would argue confusing) semantics.

If `Authorizer#removeAcls(...)` can't find anything to remove, it is supposed to return false.
If `Authorizer#removeAcls(...)` fails, it is supposed to throw an exception.
If `Authorizer#removeAcls(...)` deletes 1 or more ACLs, it is supposed to return true.

Of course we need to re-do the Authorizer interface, but for now we have to live with what we've got there.